### ISSUE TITLE
Decide on a consistent Git / Github workflow

### ISSUE THREAD

**Issue by cafca** [cafca/souma#67]
I propose that we try and consistently use a naming schema and workflow as proposed below. Separating production, development and feature branches is something that we already do, but it could be handled in a more clear and consistent manner.

http://nvie.com/posts/a-successful-git-branching-model/

Please feel free to comment (especially @cartisan @mrausch @kahartma @Francois23)

![git](https://f.cloud.github.com/assets/278463/1613922/bd0bc65e-55e1-11e3-8f7f-bd18cafadd08.jpg)

**Comment by cartisan** [comment]
Do we need a release branch? We don't have a release scheme forced upon us (by app releases the backend would have to speak e.g.) so imho feature/develop/master should suffice. The idea would be to fork branch from master for each feature, implement and test the feature, merge into develop, test with the new functionality that has accumulated there and if things are dandy merge develop into master. Means: Develop can break, but never master.
Doing that we follow the practice of continuous integration. 

(Next step would be to set up automated testing on develop and master branches to get this done automagically :) )

**Comment by cafca** [comment]
I understood the purpose of the release branch to allow for bugfixing, prettying up code, etc before a release without occupying the dev branch with this. Maybe we don't need it, I can only go by intuitions. It probably doesn't make that much of a difference as long as we stay such a small team so I don't mind going with your suggestion.

What I don't understand: Why fork feature branches from master and not from dev? Does this have an advantage?

Practically this would mean that once I finish removing the p2p-code, the rewrite branch is renamed to "develop", the old master branch gets thrown away and a new master branch is created from the new develop branch.

**Comment by cartisan** [comment]
There are two advantages and one drawback. 
Master is stable but not necessarily develop. By branching from develop one risks that the initial state for the new feature is buggy. That's not a big deal and can be discovered but running the tests before starting. What is worse however is that one could clone a state that later gets reverted in develop. Thus by merging in the feature-branch deleted features are reintroduced. We avoid this by branching from master.
The drawback is that, since master is inherently behind develop, the merging effort is potentially increased. Since it is good practice to merge develop into master constantly (continuous integration) and keep feature branches small, there shouldn't be to many conflicts to resolve manually. So I prefer branching from master.

Total agree on last part.

**Comment by cafca** [comment]
You are really fond of continuous integration! Now I get it ;) I'll close this for now.

**Comment by cafca** [comment]
@cartisan @kahartma @mrausch 

We now have a `master` and `dev` branch on both repos (glia/souma). New features are developed in branches off `master` and merged into `dev` (which is then merged into `master` again asap).
